File: ee25btech11001/MATGEO/10.7.50/codes/code.py
import numpy as np
import matplotlib.pyplot as plt

# Ellipse parameters
A = 4.0
B = 25.0
C = 100.0

# Function to compute locus y for each x
def compute_locus_py(x_vals, A, B):
    y_vals = np.zeros_like(x_vals)
    for i, x in enumerate(x_vals):
        denom = A * x**2 - B
        if denom == 0:
            y_vals[i] = np.nan
            continue
        y2 = (A * x**2) / denom
        if y2 < 0:
            y_vals[i] = np.nan
            continue
        y_vals[i] = np.sqrt(y2)
    return y_vals

# X values
x_vals = np.linspace(-6, 6, 400)
y_vals = compute_locus_py(x_vals, A, B)

# Ellipse points
theta = np.linspace(0, 2*np.pi, 400)
ellipse_x = np.sqrt(C/A) * np.cos(theta)
ellipse_y = np.sqrt(C/B) * np.sin(theta)

# Plot
plt.figure(figsize=(8,6))
plt.plot(ellipse_x, ellipse_y, label='Ellipse', color='blue')
plt.plot(x_vals, y_vals, label='Locus (upper branch)', color='red')
plt.plot(x_vals, -y_vals, color='red')
plt.axhline(0, color='black', linewidth=0.5)
plt.axvline(0, color='black', linewidth=0.5)
plt.gca().set_aspect('equal')
plt.legend()
plt.title("Ellipse and Locus")
plt.show()

File: ee25btech11001/MATGEO/10.7.50/codes/code.c
#include <math.h>
#include <stdlib.h>

// Computes y values for locus given an array of x values
// A x^2 (y^2 - 1) - B y^2 = 0  => (A x^2 - B) y^2 = A x^2
// => y^2 = (A x^2)/(A x^2 - B)
void compute_locus(double *x_vals, double *y_vals, int n, double A, double B, double C) {
    for (int i = 0; i < n; i++) {
        double x = x_vals[i];
        double denom = A * x * x - B;
        if (denom == 0) {
            y_vals[i] = NAN;  // avoid division by zero
            continue;
        }
        double y2 = (A * x * x) / denom;
        if (y2 < 0) {
            y_vals[i] = NAN;  // no real solution
            continue;
        }
        y_vals[i] = sqrt(y2);  // upper branch
    }
}

